internal/kafka/internal/handlers/kafka_promote.go

```diff
@@ -0,0 +1,76 @@
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/errors"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/handlers"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/services/authorization"
+	"github.com/gorilla/mux"
+)
+
+type kafkaPromoteHandler struct {
+	service                        services.KafkaService
+	authService                    authorization.Authorization // TODO keep or remove this depending on the other TODOs of the kafka promote handlers
+	kafkaConfig                    *config.KafkaConfig
+	kafkaPromotionValidatorFactory KafkaPromotionValidatorFactory
+}
+
+func NewKafkaPromoteHandler(service services.KafkaService, authService authorization.Authorization, kafkaConfig *config.KafkaConfig, kafkaPromotionValidatorFactory KafkaPromotionValidatorFactory) *kafkaPromoteHandler {
+	return &kafkaPromoteHandler{
+		service:                        service,
+		authService:                    authService,
+		kafkaConfig:                    kafkaConfig,
+		kafkaPromotionValidatorFactory: kafkaPromotionValidatorFactory,
+	}
+}
+
+const kafkaPromoteRouteVariableID = "id"
+
+func (h kafkaPromoteHandler) Promote(w http.ResponseWriter, r *http.Request) {
+	var kafkaPromotionRequest public.KafkaPromotionRequest
+	id := mux.Vars(r)[kafkaPromoteRouteVariableID]
+	ctx := r.Context()
+	kafkaRequest, kafkaGetError := h.service.Get(ctx, id)
+	validateKafkaFound := func() handlers.Validate {
+		return func() *errors.ServiceError {
+			return kafkaGetError
+		}
+	}
+	cfg := &handlers.HandlerConfig{
+		MarshalInto: &kafkaPromotionRequest,
+		Validate: []handlers.Validate{
+			handlers.ValidateMinLength(&kafkaPromotionRequest.DesiredKafkaBillingModel, "desired_kafka_billing_model", 1),
+			// ValidateKafkaUserFacingUpdateFields(ctx, h.authService, kafkaRequest, &kafkaUpdateReq), // TODO should some of this checks in here related to owner be performed?
+			validateKafkaFound(),
+			validateKafkaRequestToPromoteHasAPromotableActualKafkaBillingModel(kafkaRequest),
+			validateKafkaRequestToPromoteHasAPromotableStatus(kafkaRequest),
+			validateRequestedKafkaPromotionHasDifferentKafkaBillingModel(&kafkaPromotionRequest, kafkaRequest),
+			validateNoKafkaPromotionInProgress(kafkaRequest),
+			validateKafkaPromotionRequestFields(&kafkaPromotionRequest, kafkaRequest, h.service, h.kafkaConfig, h.kafkaPromotionValidatorFactory),
+		},
+		Action: func() (i interface{}, serviceError *errors.ServiceError) {
+			kafkaFieldsToUpdate := map[string]interface{}{}
+			kafkaFieldsToUpdate["desired_kafka_billing_model"] = kafkaPromotionRequest.DesiredKafkaBillingModel
+			// We clear any potential promotion details being set due to a previous
+			// Kafka promotion failure
+			if kafkaRequest.PromotionDetails != "" {
+				kafkaFieldsToUpdate["promotion_details"] = ""
+			}
+			kafkaFieldsToUpdate["promotion_state"] = dbapi.KafkaPromotionStatePromoting
+			if kafkaPromotionRequest.DesiredMarketplace != "" {
```

Comment:
Should we unset the fields instead if marketplace is not provided? That should never happen as promotion is only possible from eval where we know for sure we should never have a marketplace nor billing_cloud_account_id set beforehand as eval instances are standard kafka billing model

Reply:
Looking for some other views about this too @ziccardi @JameelB

Reply:
Yes, I believe we should: when a promotion requests arrive, we should use only the received values, regardless of any promotion attempt history

Reply:
changed